### PR TITLE
Move internal development queries into public dev branch.

### PR DESCRIPTION
**Pull request by NateD-MSFT**
In the coming months Microsoft hopes to expand the queries available in this repo by creating new driver-focused queries and by porting existing Code Analysis and Static Driver Verifier checks to CodeQL, where applicable.

This PR brings over work from an internal branch, including:

- Early .qll libraries for reasoning about WDM and KMDF drivers
- A port of the DeviceInitApi Static Driver Verifier rule, and associated test drivers
- Creation of a NOTICE.md file to account for queries and libraries included from the main CodeQL repo.

### CODE REVIEW

**Review by t-abtamene — APPROVED** [review]

File: codeql/windows-drivers/queries/Windows/wdk/kmdf/DeviceInitApi/test/pass/README.md
Comment (lauragonzalezzz) [inline codeql/windows-drivers/queries/Windows/wdk/kmdf/DeviceInitApi/test/pass/README.md:1]: Adjust to pertain to CodeQL, WDDST.

File: codeql/windows-drivers/queries/Windows/wdk/kmdf/DeviceInitApi/test/pass/library/fail_library1.c
Comment (lauragonzalezzz) [inline codeql/windows-drivers/queries/Windows/wdk/kmdf/DeviceInitApi/test/pass/library/fail_library1.c:1]: Are the fail drivers in the PASS library the same as the ones in the FAIL library?
Whats the difference between the Pass/Fail directories if so?

File: NOTICE.md
Comment (NateD-MSFT) [reply NOTICE.md:1]: It's common practice to include third-party licenses in a separate file from your own license.  For example: https://github.com/microsoft/terminal/blob/main/NOTICE.md

File: codeql/windows-drivers/queries/Windows/wdk/kmdf/DeviceInitApi/DeviceInitApi.ql
Comment (NateD-MSFT) [reply codeql/windows-drivers/queries/Windows/wdk/kmdf/DeviceInitApi/DeviceInitApi.ql:1]: Will do.

File: codeql/windows-drivers/queries/suites/windows_driver_recommended.qls
Comment (NateD-MSFT) [reply codeql/windows-drivers/queries/suites/windows_driver_recommended.qls:4]: Since this is the recommended file (which includes the MustFix queries as an import, plus the recommended ones), how does "recommended and required queries for Windows Drivers"sound?

File: codeql/windows-drivers/queries/Windows/wdk/wdm/SAL.qll
Comment (NateD-MSFT) [reply codeql/windows-drivers/queries/Windows/wdk/wdm/SAL.qll:6]: Valid was a poor choice of words.  The default SAL library in the CodeQL repo does not refer to driverspecs.h as it is a header only meaningful to Windows drivers.  In the past we have gotten some pushback when asking if we should bring Windows driver-specific code into the main CodeQL repo, but I can open an issue to get more feedback from them on this specific case.  

For TODOs, do you know if we generally publicize AzDO work item numbers in public code? That strikes me as a little odd, but I could be overthinking it.

File: codeql/windows-drivers/queries/Windows/wdk/wdm/SAL.qll
Comment (NateD-MSFT) [reply codeql/windows-drivers/queries/Windows/wdk/wdm/SAL.qll:54]: I'd actually like to push back against this change for two reasons:

1. This file is taken from the main CodeQL repo - it's only duplicated here because we need to include "driverspecs.h" as a source of annotations.  So I would prefer not to re-write comments taken directly from the main repo.  
2. The classes are the CodeQL representations of the SAL macros.  Looking at other CodeQL core libraries, this style of comment (where a class is simply described as what it is in the code being analyzed, rather than as "getting" that thing) is common.  See for example: https://github.com/github/codeql/blob/main/cpp/ql/lib/semmle/code/cpp/Union.qll

File: NOTICE.md
Comment (NateD-MSFT) [reply NOTICE.md:2]: I'm not certain if we will eventually include non-WDK content or not, but for library organization I do prefer to have a folder/namespace that is for drivers and not just general Windows code so that its purpose is clear.  We could rename it to "drivers" instead of "wdk", perhaps?

Re: KMDF/WDM: at the time I made this I was mostly following the SDV style (which includes WDM, KMDF, NDIS, and storport-specific rules divided into those folders.)  I doubt that we plan to write UMDF queries in the immediate future, so I don't know that we should add a "UMDF" folder.

Thoughts?

File: codeql/windows-drivers/queries/Windows/wdk/wdm/WdmDrivers.qll
Comment (NateD-MSFT) [reply codeql/windows-drivers/queries/Windows/wdk/wdm/WdmDrivers.qll:156]: WDM.h - I will add this detail to the comment.

File: codeql/windows-drivers/queries/Windows/wdk/kmdf/DeviceInitApi/DeviceInitApi.ql
Comment (NateD-MSFT) [reply codeql/windows-drivers/queries/Windows/wdk/kmdf/DeviceInitApi/DeviceInitApi.ql:6]: Added details about why this causes instability (the framework takes ownership of the object).  This is not a security bug and as such is tagged correctness rather than security.

File: codeql/windows-drivers/queries/Windows/wdk/kmdf/DeviceInitApi/test/pass/library/fail_library1.c
Comment (NateD-MSFT) [reply codeql/windows-drivers/queries/Windows/wdk/kmdf/DeviceInitApi/test/pass/library/fail_library1.c:1]: The fail folder has an injected defect for this query, while the pass folder does not.  I did not rename the individual files in the pass folder in part because these full drivers are stopgaps until we implement Abenezer+Jakob's code snippet+template test case style.

File: codeql/windows-drivers/queries/Windows/wdk/wdm/InconsistentDispatchAnnotations.ql
Comment (NateD-MSFT) [reply codeql/windows-drivers/queries/Windows/wdk/wdm/InconsistentDispatchAnnotations.ql:22]: This would require some extra refactoring/library functions to properly print what the driver was assigned to so I'd prefer not to block on that for the moment - for now I've expanded the message to be more clear and indicate what the mistaken annotation is and what to review.

File: NOTICE.md
Comment (lauragonzalezzz) [reply NOTICE.md:2]: I don't think we need so many levels in the repo in general, and the distinction between these is unclear. The purpose of the repo is for CodeQL Windows driver development and analysis, and that is mostly inherent in the name.
Level 1: We're probably not going to add any other static analysis tools to this repo in the near future so we can probably remove the "**codeql**" level without any issue, and add it later if there ever comes a need.
Level 2: We have "**codeql-queries**" and "**windows-drivers/queries**" which is essentially "**general queries**" and "**driver specific queries**" (note that using slashes in directory names is not a good practice and introduces further confusion since the slash denotes hierarchy and not a separation of topic; use dashes if you must but they are probably not necessary here)
Level 3: (Under "windows-drivers/queries") we have "**Likely Bugs**" "**Security/Cryptography**" "**Windows/wdk**" "**microsoft/code/cpp/commons**" and "**suites**" which is extremely confusing as there is no clear organization or distinction between these. Is Security/Cryptography a category of rules? If so, wouldn't that belong in "**Suites**"? Is "microsoft/code/cpp/commons" not part of "Windows/wdk"? Perhaps these should move under a directory called "**libraries**" since they are not queries?
Level 4: (Under "Windows/wdk") Is there a material difference in how you approach a KMDF driver query versus a WDM query or a logical way of organizing these queries? In my opinion, the purpose of the distinction is unclear because WDM should be a parent to KMDF, but if we write the queries exactly the same whether they are WDM or WDM/KMDF, then what is the need to make the separation? For example, if we have sibling categories "Apples" and "Green Apples." Unless we're doing something special on "Green Apples" then why separate them from the other apples? That's why I suggested adding the UMDF directory. Because having parent directory "Apples" with sub directories "Green Apples" and "Red Apples" makes more sense to me since we treat "Green Apples" (KMDF queries) and "Red Apples" (UMDF queries) differently.

File: codeql/windows-drivers/queries/Windows/wdk/wdm/SAL.qll
Comment (lauragonzalezzz) [reply codeql/windows-drivers/queries/Windows/wdk/wdm/SAL.qll:6]: If you are concerned about the potential security risk of publishing a task number then I'd advise you to ask the appropriate teams, otherwise I think it is wise to be able to track the items that aren't finished yet and provide a small indication of such to the public.

File: codeql/windows-drivers/queries/Windows/wdk/wdm/SAL.qll
Comment (lauragonzalezzz) [reply codeql/windows-drivers/queries/Windows/wdk/wdm/SAL.qll:54]: If this file is coming directly from the CodeQL repo then why is it copied here and not referenced from the mirrored codeql repo located at Windows-Driver-Developer-Supplemental-Tools/codeql/codeql-queries @ e178626 ?

File: codeql/windows-drivers/queries/Windows/wdk/kmdf/DeviceInitApi/test/pass/library/fail_library1.c
Comment (lauragonzalezzz) [reply codeql/windows-drivers/queries/Windows/wdk/kmdf/DeviceInitApi/test/pass/library/fail_library1.c:1]: We will need to discuss how to separate test code from queries code in the hierarchy to make it easier to check in for release. We dont want customers accidentally downloading a bunch of test code. But I see your point and agree.

File: codeql/windows-drivers/queries/Windows/wdk/wdm/SAL.qll
Comment (NateD-MSFT) [reply codeql/windows-drivers/queries/Windows/wdk/wdm/SAL.qll:6]: TODO: Add a TODO: to the comment block to be more clear.

File: codeql/windows-drivers/queries/Windows/wdk/wdm/SAL.qll
Comment (NateD-MSFT) [reply codeql/windows-drivers/queries/Windows/wdk/wdm/SAL.qll:54]: Because it was edited to add driverspecs.h to the list and is not a 1:1 copy.  I tried originally just to extend this class in a file we own, but unfortunately CodeQL inheritance doesn't work that way, so for now we need to have our own copy with our edits.

File: NOTICE.md
Comment (NateD-MSFT) [reply NOTICE.md:2]: I'm working to reorganize this per your guidelines above, but just to answer a few of your questions:
- The "Level 3" folders and organization are largely because we took several queries and libraries from the internal Microsoft CodeQL repo (Raul and Josh's repo) and were moved that way to avoid breaking any dependencies and to maintain organizational consistency with as they were in that query.  We can move them around, it just may be slightly more work in the future to update them with any changes from Raul/Josh's internal repo.
- Although it's true that WDM can in some ways be thought of a "parent" of KMDF, when validating driver behavior (especially looking forward to porting SDV rules), having a separation in queries and libraries is useful and even necessary.  Behavior that is correct in a WDM context may not be correct in a KMDF context and vice-versa.  There may be some libraries and queries that are universally applicable (particularly in the CA space) but I would prefer to leave "wdm" and "kmdf" as siblings as opposed to a parent/child relationship, with global queries in their parent folder.  If desired we could have a structure where "windows/drivers" is the top level, "wdm" and "wdf" are the children, and "kmdf" and "umdf" are the children of "wdf".

**Comment by NateD-MSFT** [comment]
FYI: After the reorganization commit, this is how the repo layout looks to me.  Thoughts? @lauragonzalezzz 
image

File: src/drivers/kmdf/queries/DeviceInitApi/test/pass/README.md
Comment (lauragonzalezzz) [inline src/drivers/kmdf/queries/DeviceInitApi/test/pass/README.md:3]: Should this say CodeQL?

File: src/drivers/kmdf/queries/DeviceInitApi/test/pass/library/fail_library1.c
Comment (lauragonzalezzz) [inline src/drivers/kmdf/queries/DeviceInitApi/test/pass/library/fail_library1.c:12]: CodeQL?

File: src/drivers/libraries/SAL.qll
Comment (lauragonzalezzz) [inline src/drivers/libraries/SAL.qll:25]: Are we supporting anything before Windows 7?

Comment (NateD-MSFT) [reply src/drivers/libraries/SAL.qll:25]: I doubt it (and I'm actually unsure if SAL was available for public use before Windows 7 - at the very least Win7 was where we first introduced PreFAST for Drivers, which used SAL for driver analysis).  In any event this code is just taken 1:1 from the main CodeQL repo.

File: codeql/windows-drivers/queries/Windows/wdk/wdm/SAL.qll
Comment (lauragonzalezzz) [reply codeql/windows-drivers/queries/Windows/wdk/wdm/SAL.qll]: If it is a straight copy then I can understand not wanting to change it at all, but if you are making changes to it anyway then why not update it to be clearer? I doubt we will be RI-ing anyway.
I'm not going to block on this but I believe it is worth considering.

File: src/drivers/kmdf/queries/DeviceInitApi/test/fail/README.md
Comment (lauragonzalezzz) [inline src/drivers/kmdf/queries/DeviceInitApi/test/fail/README.md:11]: Please make a todo to remove all references to SDV, including in file names and function names. They are throughout the code and unnecessarily complicate it, but I will not block on that.

Comment (NateD-MSFT) [reply src/drivers/kmdf/queries/DeviceInitApi/test/fail/README.md:11]: Thanks.  I will add a TODO, but FWIW my expectation is that once Abenezer merges his PR I'll create a new PR removing these full-size drivers and replacing them with code snippets that work with the testing script he and Jakob put together.

**Comment by NateD-MSFT** [comment]
Laura approved over Teams.  Squashing and merging!

File: src/drivers/libraries/SAL.qll
Comment (JakobL-MSFT) [reply src/drivers/libraries/SAL.qll:25]: We should not support Windows 8.1, Windows 8, or Windows 7.  This to follow the style of the newest WDK.

File: src/drivers/kmdf/queries/DeviceInitApi/test/fail/README.md
Comment (JakobL-MSFT) [reply src/drivers/kmdf/queries/DeviceInitApi/test/fail/README.md:11]: I am very much looking forward to Nate leveraging the snippet unit test framework.  That will make unit test oh so much more compact and succinct.

File: codeql/windows-drivers/queries/Windows/wdk/wdm/SAL.qll
Comment (JakobL-MSFT) [reply codeql/windows-drivers/queries/Windows/wdk/wdm/SAL.qll]: It is always ideal to not have to copy files.  But sometimes it is a reasonable pragmatic choice to copy.    (Typically, that then comes and bites you about 5 years later :-) )

In the case where you copy a file I am personally a fan of "minimizing the diff".  So, make changes as clean as possible for people in future running "windiff orig patched".  It is also reasonable to write a comment providing this context in the file"/* This file copied from x\y\z and only following line(s) was edited */".  That line will show up yellow of course, but the intension will be immediately clear.

My 2 cents.

**Pull request merged**
